### PR TITLE
(#1694605) Backport/fixes for issues found by LGTM on RHEL 7

### PR DESCRIPTION
**Pull request by mrc0mmand**
This PR attempts to fix errors & warnings found by LGTM. Recommendations are currently ignored, as it would make this PR unnecessarily big.

**Note:** some of these patches are not even in upstream (mostly because the particular code was completely rewritten in upstream).

Unresolved errors/warnings:
 - **[error]** `src/shared/pager.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/fe7b6b6c19424e12d6fb721e94422a5ff28b7125/files/src/shared/pager.c?sort=name&dir=ASC&mode=heatmap#x437173520f29364c:1) - pager-related stuff has been significantly refactored in upstream, needs further investigation
 - ~~*[ignored]* `src/libsystemd/sd-bus/bus-bloom.h` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/libsystemd/sd-bus/bus-bloom.c?sort=name&dir=ASC&mode=heatmap) - dropped from upstream along with kdbus~~
 - `src/shared/strv.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/shared/strv.c?sort=name&dir=ASC&mode=heatmap#x90c37d31240b843a:1) - `strv_join_quoted` was dropped from upstream (replaced by a different approach) without resolving this issue; would require some, possibly substantial, refactoring
 - ~~*[ignored]* `src/bootchart/*` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/bootchart/bootchart.c#x9c7e30d234d48467:1) - dropped from upstream (systemd/systemd@232c84b2d22f2d96982b3c0390d29498bb430835)~~
 - `src/core/execute.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/core/execute.c?sort=name&dir=ASC&mode=heatmap#xbfa2ae16fedaa55f:1) - `process_execute` has been refactored to support longer process names (systemd/systemd@9bfaffd5a9c138b0bbdfb452b3f4435746f2d18b), not sure if it's worth backporting
 - ~~*[ignored]* `src/gudev/*` - dropped from upstream (systemd/systemd@2375607039517c88df51ef16ddbb624ec1c10654)~~
 - `src/shared/fw-util.c` [[link]](https://lgtm.com/projects/g/lnykryn/systemd-rhel/snapshot/f5844bf3f34f5e468c9c28487c37de607d85fab1/files/src/shared/fw-util.c#x62def701dfc13e50:1) - renamed to `src/shared/firewall-util.c`, however, not yet fixed in upstream

The *[ignored]* rules should *hopefully* take effect once this PR is merged.

### CODE REVIEW

**Comment by lnykryn** [comment]
This pull request **fixes 13 alerts** when merging 4df22c749bd91260f314bb94ad38ad3e735d368f into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-627246f7bbd1809db84daf05eea60f337f32a9e5)

**fixed alerts:**

* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 15 alerts** when merging 73386a79dec0cfd5a9341f7b4c95e82fb98da1a0 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-7fd5027e09ab908560398da1bb228b78e183331e)

**fixed alerts:**

* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 2 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 23 alerts** when merging 368d4f48e517f7765165a56f7a8b11627f334fa8 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-c90f6102cff5c39de985530a061ca8007a2e1c82)

**fixed alerts:**

* 7 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **introduces 1 alert** and **fixes 25** when merging 300a5ddc3f3f123841929288ad6a7c1e0648f5c9 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-2ffcf7979bde597b4eb569164fbdd71c9f517e63)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 8 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 26 alerts** when merging 5a5a6768aad8171238913cbc8db0a9f738c06b7a into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-08935a601b81d42e9228fe2bfa17363fc0b8afa3)

**fixed alerts:**

* 9 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 28 alerts** when merging ba2518d9a6ce01a303c89b4c48cf84f1b822c6dd into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-64abe63b5cce3a4c4bac0b92c8cfe7691d4a5b79)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 35 alerts** when merging 70c92469496a831f01ef953442ac8b9c89a050e0 into a781c22627a108d2fdcfaf428f6fefca122636a9 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-3633f76dbe5c6b0fa365769cfff97ac881cb10d8)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
Even though this PR doesn't fix all identified warnings, I'd review and merge it in its current state, as stuff's getting too hard to follow.

**Comment by jsynacek** [comment]
IMHO all the commits need to include "Resolves:" or "Related:", otherwise it won't be possible to merge them downstream.

**Comment by mrc0mmand** [comment]
@jsynacek thanks for pointing it out, it completely slipped my mind. Should be fixed now.

**Comment by lnykryn** [comment]
This pull request **fixes 35 alerts** when merging 3302ab3e54b53f30bce3ab4a52cc6db62f0edb02 into 25112918d6b54fc1136dde2eef24d6c2d33e55b3 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-914c9183086959ef8bb15ee1b24c7cffa36dfd51)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 4 for Potentially overflowing call to snprintf
* 3 for Use of potentially dangerous function
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/journal/journal-verify.c
Comment (jsynacek) [inline src/journal/journal-verify.c:77]: The original commit used the c-style comment, not the c++-style one. Even though it's really a triviality, I'd say that we should do the same here, as systemd uses the c-style comments everywhere.

File: src/sysv-generator/sysv-generator.c
Comment (jsynacek) [inline src/sysv-generator/sysv-generator.c:571]: Were these changes cherry picked from somewhere? If yes, please mention the patch in the commit message. If not, please add "rhel-only" at the end of the commit message.

Also, use the c-style comment, as mentioned in the other patch.

**Review by jsynacek — CHANGES_REQUESTED** [review]
I've found more "problems". Sorry for not finding them earlier and making you rebase again.

File: src/journal/journal-verify.c
Comment (mrc0mmand) [reply src/journal/journal-verify.c:77]: This is an unfortunate behavior of LGTM - it simply doesn't recognize suppression rules in C++-style comments. I had to do a similar change in upstream as well (see https://github.com/systemd/systemd/commit/1a862e21511918f4f7a27a119624e9629056356b) as the rule was ignored within a C++-style comment.

Excerpt from the [LGTM documentation](https://lgtm.com/help/lgtm/alert-suppression#c-c++):
```
The syntax is language-dependent:

    "// lgtm..." for C/C++, C#, Java and JavaScript
    "* lgtm..." for COBOL
    "# lgtm..." for Python
```

File: src/sysv-generator/sysv-generator.c
Comment (mrc0mmand) [reply src/sysv-generator/sysv-generator.c:571]: Will add "rhel-only" to the commit, as the error-handling part of the `load_sysv()` function was significantly rewritten. Also, for the comment-style, the same explanation applies as above.

**Comment by lnykryn** [comment]
This pull request **fixes 7 alerts** when merging 72b47a1d57adf79528dca9dbd9a0b152d181f27e into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-8c6d1bdf91a68e7ba2a1489f4164b6b2682e054a)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by lnykryn** [comment]
This pull request **fixes 7 alerts** when merging 72b47a1d57adf79528dca9dbd9a0b152d181f27e into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-8c6d1bdf91a68e7ba2a1489f4164b6b2682e054a)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
Ugh, the LGTM analysis now keeps failing due to compilation error:

```
[2019-04-04 08:45:00] [build] make --no-print-directory all-recursive
[2019-04-04 08:45:00] [build] Making all in .
[2019-04-04 08:45:00] [build]   CC       src/cgls/cgls.o
[2019-04-04 08:45:01] [build] In file included from ./src/shared/util.h:92,
[2019-04-04 08:45:01] [build]                  from ./src/shared/cgroup-show.h:26,
[2019-04-04 08:45:01] [build]                  from src/cgls/cgls.c:29:
[2019-04-04 08:45:01] [build] ./src/shared/missing.h:989:19: error: static declaration of ‘renameat2’ follows non-static declaration
[2019-04-04 08:45:01] [build]  static inline int renameat2(int oldfd, const char *oldname, int newfd, const char *newname, unsigned flags) {
[2019-04-04 08:45:01] [build]                    ^~~~~~~~~
[2019-04-04 08:45:01] [build] In file included from src/cgls/cgls.c:23:
[2019-04-04 08:45:01] [build] /usr/x86_64-linux-gnu/include/stdio.h:164:12: note: previous declaration of ‘renameat2’ was here
[2019-04-04 08:45:01] [build]  extern int renameat2 (int __oldfd, const char *__old, int __newfd,
[2019-04-04 08:45:01] [build]             ^~~~~~~~~
[2019-04-04 08:45:02] [build] make[2]: *** [Makefile:15186: src/cgls/cgls.o] Error 1
[2019-04-04 08:45:02] [build] make[1]: *** [Makefile:19815: all-recursive] Error 1
[2019-04-04 08:45:02] [build] make: *** [Makefile:10424: all] Error 2
```

**Comment by mrc0mmand** [comment]
This seems to be a variation of https://github.com/systemd/systemd/issues/8099 fixed by https://github.com/systemd/systemd/commit/5187dd2c403caf92d09f3491e41f1ceb3f10491f, we may need to backport it, somehow...

**Comment by lnykryn** [comment]
This pull request **fixes 7 alerts** when merging 38b4fbc5d310e3979804ec3a2f94850eb4e64e91 into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-bb57825c6c6b527ac4ef9d5e3e08fea433268c6b)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
As the LGTM infrastructure got upgraded, mainly with newer glibc, I had to fix a few other issues:

1) dd5e0d980490fcf394faf3175a13ff555949dc10
This fixes issue mentioned in https://github.com/lnykryn/systemd-rhel/pull/311#issuecomment-479818375. To be somewhat consistent across the `missing.h` file I added the same *workaround* for other syscalls as well, not only the `renameat2`, which was causing issues

2) 3fec580890591c7d4b80efd0ebb2cec7120191dc
Newer glibc stops including `sys/sysmacros.h` from the `sys/types.h`, which causes compilation issues in code using `major()` and `minor()` functions, e.g.:

```
[2019-04-04 09:32:27] [build] /usr/bin/ld: ./.libs/libsystemd-shared.a(libsystemd_shared_la-util.o): in function `get_ctty_devnr':
[2019-04-04 09:32:27] [build] util.c:(.text.get_ctty_devnr+0x8f): undefined reference to `major'
[2019-04-04 09:32:27] [build] /usr/bin/ld: util.c:(.text.get_ctty_devnr+0x12e): undefined reference to `minor'
[2019-04-04 09:32:27] [build] collect2: error: ld returned 1 exit status
```

Let's add it explicitly to files which need it.


With these two additional commits the LGTM analysis should work once again.

**Comment by lnykryn** [comment]
This pull request **fixes 7 alerts** when merging 3fec580890591c7d4b80efd0ebb2cec7120191dc into ac46d01c5f6a211bbbbb43e20f63ecae2549da20 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-6afc2692204bf50d3c557fd8d86dcbb9cbcb24f7)

**fixed alerts:**

* 6 for Missing variable declaration
* 1 for Unneeded defensive code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
The current LGTM fail is caused by the last two commits not being present in the staging branch, should be fixed once this PR is merged.

image

**Comment by lnykryn** [comment]
This pull request **fixes 38 alerts** when merging d75397fef3651631869653cab7165ddcd3dfc6b2 into dd004bcc058dbb7c66b8eb1a728732ccfce746c7 - [view on LGTM.com](https://lgtm.com/projects/g/lnykryn/systemd-rhel/rev/pr-1233127520b4663870f526aa100c5e4bd6daf890)

**fixed alerts:**

* 11 for Comparison result is always the same
* 6 for Missing variable declaration
* 6 for Assignment where comparison was intended
* 6 for Use of potentially dangerous function
* 4 for Potentially overflowing call to snprintf
* 3 for Implicit downcast from bitfield
* 1 for Unneeded defensive code
* 1 for Ambiguously signed bit\-field member

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
Closed in favor of https://github.com/systemd-rhel/rhel-7/pull/2

**Pull request closed without merging**
